src/rt/sched/behaviourcore.h

```diff
@@ -735,178 +740,226 @@ namespace verona::rt
+        auto* first_body = bodies[first_body_index];
+        auto* new_slot = std::get<2>(chain_info[i]);
+
         auto prev_slot =
-          cown->last_slot.exchange(curr_slot, std::memory_order_acq_rel);
+          cown->last_slot.exchange(new_slot, std::memory_order_acq_rel);
+
         yield();
 
         if (prev_slot == nullptr)
         {
-          if (curr_slot->is_read_only())
-          {
-            yield();
-            bool first_reader = cown->read_ref_count.add_read();
-            Logging::cout() << "Reader at head of queue and got the cown "
-                            << *curr_slot << Logging::endl;
-            yield();
-            curr_slot->set_read_available();
-            ec[std::get<0>(indexes[first_chain_index])]++;
-            yield();
-            acquire_with_transfer(cown, transfer_count, 1 + first_reader);
-            continue;
-          }
-
-          yield();
-
-          acquire_with_transfer(cown, transfer_count, 1);
-
-          yield();
-
-          if (cown->read_ref_count.try_write())
-          {
-            Logging::cout() << " Writer at head of queue and got the cown "
-                            << *curr_slot << Logging::endl;
-            ec[std::get<0>(indexes[first_chain_index])]++;
-            yield();
-            continue;
-          }
-
-          Logging::cout() << " Writer waiting for previous readers cown "
-                          << *curr_slot << Logging::endl;
-          yield();
-          cown->next_writer = body;
+          std::get<4>(chain_info[i]) = true;
           continue;
         }
 
-        // Release any transferred count
-        acquire_with_transfer(cown, transfer_count, 0);
-
-        yield();
-        Logging::cout() << " Someone in queue cown " << *curr_slot
-                        << " previous " << *prev_slot << Logging::endl;
-
         while (prev_slot->is_wait_2pl())
         {
           Systematic::yield_until(
             [prev_slot]() { return !prev_slot->is_wait_2pl(); });
           Aal::pause();
         }
 
-        if (curr_slot->is_read_only())
+        if (new_slot->is_read_only())
         {
-          if (prev_slot->set_next_slot_reader(curr_slot))
+          if (prev_slot->set_next_slot_reader(new_slot))
           {
             Logging::cout()
               << " Previous slot is a writer or blocked reader cown "
-              << *curr_slot << Logging::endl;
+              << *new_slot << Logging::endl;
             yield();
             continue;
           }
 
           yield();
           bool first_reader = cown->read_ref_count.add_read();
-          Logging::cout() << " Reader got the cown " << *curr_slot
+          Logging::cout() << " Reader got the cown " << *new_slot
                           << Logging::endl;
           yield();
-          curr_slot->set_read_available();
-          ec[std::get<0>(indexes[first_chain_index])]++;
+          new_slot->set_read_available();
```

Comment:
This will not be correct in the multi-schedule case as we will need to check if the chain contains only reads.  It might be good to calculate this in the prepare phase, as it means we can quickly pass on the read available token.  It is probably just needs another bool for `all_reads`.